### PR TITLE
CVSB-163

### PR DESCRIPTION
**Pull request by mgherceanu**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -93,7 +93,7 @@
     "karma-sourcemap-loader": "^0.3.7",
     "karma-webpack": "^2.0.3",
     "null-loader": "^0.1.1",
-    "protractor": "^5.1.1",
+    "protractor": "^5.4.2",
```
Comment (danieliordache) [inline package.json]: why this update ?

File: src/app/app.component.ts [hunk 2]
```diff
@@ -22,6 +22,15 @@ export class MyApp {
       statusBar.overlaysWebView(true);
       statusBar.styleLightContent();
 
+      // easter egg
+      if (AppConfig.IS_PRODUCTION == 'true') {
+        localStorage.setItem('easterEgg', 'false');
+        localStorage.setItem('caching', 'true');
+      } else {
+        localStorage.setItem('easterEgg', 'true');
```
Comment (danieliordache) [inline src/app/app.component.ts]: use an enum variable

Comment (danieliordache) [reply src/app/app.component.ts]: for easterEgg and caching

File: src/models/visit/activities.model.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+export interface ActivitiesModel {
```
Comment (danieliordache) [inline src/models/visit/activities.model.ts]: rename to ActivityModel

File: src/pages/testing/test-creation/test-cancel/test-cancel.ts [hunk 4]
```diff
@@ -34,6 +34,7 @@ export class TestCancelPage {
             text: 'Submit',
             cssClass: 'danger-action-button',
             handler: () => {
+              console.log(this.testData)
```
Comment (danieliordache) [inline src/pages/testing/test-creation/test-cancel/test-cancel.ts]: remove console.log

File: src/pages/visit/end-visit-confirm/end-visit-confirm.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { Component } from '@angular/core';
+import { IonicPage, NavController, NavParams } from 'ionic-angular';
+
+@IonicPage()
+@Component({
+  selector: 'page-end-visit-confirm',
+  templateUrl: 'end-visit-confirm.html',
+})
+export class EndVisitConfirmPage {
+  atfName: string;
```
Comment (danieliordache) [inline src/pages/visit/end-visit-confirm/end-visit-confirm.ts]: change atf to testStation

File: src/pages/visit/visit-timeline/visit-timeline.ts [hunk 6]
```diff
@@ -51,4 +60,32 @@ export class VisitTimelinePage implements OnInit {
...
+          role: 'cancel'
+        },
+        {
+          text: APP_STRINGS.CONFIRM,
+          handler: () => {
+            this.loading.present();
+            this.visitService.endVisit(this.visit.id).subscribe(
+              () => {
+                this.visitService.visit.endTime = new Date().toISOString();
+                this.storageService.delete('visit');
```
Comment (danieliordache) [inline src/pages/visit/visit-timeline/visit-timeline.ts]: use STORAGE[key] from app.enums

Comment (danieliordache) [inline src/pages/visit/visit-timeline/visit-timeline.ts]: use the variable from app.enums

File: src/pages/visit/visit-timeline/visit-timeline.ts [hunk 7]
```diff
@@ -51,4 +60,32 @@ export class VisitTimelinePage implements OnInit {
...
+                this.storageService.delete('state');
+                this.visitService.visit = {} as VisitModel;
+                this.loading.dismissAll();
+                this.navCtrl.push('EndVisitConfirmPage', {atfName: this.visit.testStationName});
```
Comment (danieliordache) [inline src/pages/visit/visit-timeline/visit-timeline.ts]: rename atfName to testStationName

File: src/providers/visit/visit.service.ts [hunk 8]
```diff
@@ -21,13 +29,26 @@ export class VisitService {
...
     this.visit.tests = [];
+    if (id) this.visit.id = id;
     this.updateVisit();
     return this.visit;
   }
 
-  endVisit() {
-    this.visit.endTime = new Date().toISOString();
-    this.updateVisit();
+  startVisit(testStation) {
```
Comment (danieliordache) [inline src/providers/visit/visit.service.ts]: add return type

File: src/providers/visit/visit.service.ts [hunk 9]
```diff
@@ -21,13 +29,26 @@ export class VisitService {
...
+      testStationPNumber: testStation.testStationPNumber,
+      testStationEmail: testStation.testStationEmails[0],
+      testStationType: testStation.testStationType,
+      testerName: 'Maria Ciobanu',
+      testerStaffId: '2019'
+    };
+    return this.httpService.startVisit(activities);
+  }
+
+  endVisit(id: string) {
```
Comment (danieliordache) [inline src/providers/visit/visit.service.ts]: add return type

File: src/app/app.component.ts (hunk 2 above)
Comment (mgherceanu) [reply src/app/app.component.ts]: done

File: src/models/visit/activities.model.ts (hunk 3 above)
Comment (mgherceanu) [reply src/models/visit/activities.model.ts]: done

File: src/pages/testing/test-creation/test-cancel/test-cancel.ts (hunk 4 above)
Comment (mgherceanu) [reply src/pages/testing/test-creation/test-cancel/test-cancel.ts]: done

File: src/pages/visit/end-visit-confirm/end-visit-confirm.ts (hunk 5 above)
Comment (mgherceanu) [reply src/pages/visit/end-visit-confirm/end-visit-confirm.ts]: done

File: src/pages/visit/visit-timeline/visit-timeline.ts (hunk 6 above)
Comment (mgherceanu) [reply src/pages/visit/visit-timeline/visit-timeline.ts]: done

Comment (mgherceanu) [reply src/pages/visit/visit-timeline/visit-timeline.ts]: done

File: src/pages/visit/visit-timeline/visit-timeline.ts (hunk 7 above)
Comment (mgherceanu) [reply src/pages/visit/visit-timeline/visit-timeline.ts]: done

File: src/providers/visit/visit.service.ts (hunk 9 above)
Comment (mgherceanu) [reply src/providers/visit/visit.service.ts]: done

File: package.json (hunk 1 above)
Comment (mgherceanu) [reply package.json]: i used npm install and npm audit fix, when checking for vulnerabilities, so this got updated. 
npm uses the tilde (~) and caret (^) to designate which patch and minor versions to use respectively.

So if you see ~1.0.2 it means to install version 1.0.2 or the latest patch version such as 1.0.4. If you see ^1.0.2 it means to install version 1.0.2 or the latest minor or patch version such as 1.1.0.

there was an update package.json dependencies msgin my ide also.

File: src/pages/visit/visit-timeline/visit-timeline.ts [hunk 10]
```diff
@@ -66,4 +72,32 @@ export class VisitTimelinePage implements OnInit {
...
+          text: APP_STRINGS.CANCEL,
+          role: 'cancel'
+        },
+        {
+          text: APP_STRINGS.CONFIRM,
+          handler: () => {
+            this.loading.present();
+            this.visitService.endVisit(this.visit.id).subscribe(
+              () => {
+                this.visitService.visit.endTime = new Date().toISOString();
```
Comment (danieliordache) [inline src/pages/visit/visit-timeline/visit-timeline.ts]: remove it at is unused for display

Comment (mgherceanu) [reply src/pages/visit/visit-timeline/visit-timeline.ts]: done

**Pull request merged**